storage/http_client.go

```diff
@@ -975,6 +986,195 @@ func (c *httpStorageClient) NewRangeReader(ctx context.Context, params *newRange
 	}, nil
 }
 
+func (c *httpStorageClient) newRangeReaderJSON(ctx context.Context, params *newRangeReaderParams, opts ...storageOption) (r *Reader, err error) {
+	s := callSettings(c.settings, opts...)
+	call := c.raw.Objects.Get(params.bucket, params.object)
+
+	setClientHeader(call.Header())
+	call.Context(ctx)
+	call.Projection("full")
+
+	if s.userProject != "" {
+		call.UserProject(s.userProject)
+	}
+
+	if params.readCompressed {
+		call.Header().Set("Accept-Encoding", "gzip")
+	}
+	if err := setEncryptionHeaders(call.Header(), params.encryptionKey, false); err != nil {
+		return nil, err
+	}
+
+	// Define a function that initiates a Read with offset and length, assuming we
+	// have already read seen bytes.
+	reopen := func(seen int64) (*http.Response, error) {
+		// If the context has already expired, return immediately without making a
+		// call.
+		if err := ctx.Err(); err != nil {
+			return nil, err
+		}
+		start := params.offset + seen
+		if params.length < 0 && start < 0 {
+			call.Header().Set("Range", fmt.Sprintf("bytes=%d", start))
+		} else if params.length < 0 && start > 0 {
+			call.Header().Set("Range", fmt.Sprintf("bytes=%d-", start))
+		} else if params.length > 0 {
+			// The end character isn't affected by how many bytes we've seen.
+			call.Header().Set("Range", fmt.Sprintf("bytes=%d-%d", start, params.offset+params.length-1))
+		}
+		// We wait to assign conditions here because the generation number can change in between reopen() runs.
+		if err := applyConds("NewReader", params.gen, params.conds, call); err != nil {
```

Comment:
We should probably add a test to verify that preconditions including ifGenerationNotMatch are passed through as expected

Reply:
Added an integration test for the conditions on reader